src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -0,0 +1,422 @@
+
+            /// <summary>
+            /// The <see cref="IVsWindowFrame"/> of the active window. This may be null if we're in the middle of construction and
+            /// we haven't been given it yet.
+            /// </summary>
+            private IVsWindowFrame? _windowFrame;
+
+            private string? _windowFrameMessageToShow = null;
+            private ImageMoniker _windowFrameImageMonikerToShow = default;
+            private string? _currentWindowFrameMessage = null;
+            private ImageMoniker _currentWindowFrameImageMoniker = default;
+            private IVsInfoBarUIElement? _currentWindowFrameInfoBarElement = null;
+
+            public OpenSourceGeneratedFile(SourceGeneratedFileManager fileManager, ITextBuffer textBuffer, Workspace workspace, ProjectId projectId, string generatorTypeName, string generatedSourceHintName, IThreadingContext threadingContext)
+                : base(threadingContext, assertIsForeground: true)
+            {
+                _fileManager = fileManager;
+                _textBuffer = textBuffer;
+                _workspace = workspace;
+                _projectId = projectId;
+                _generatorTypeName = generatorTypeName;
+                _generatedSourceHintName = generatedSourceHintName;
+
+                // We'll create a read-only region for the file, but it'll be a dynamic region we can temporarily suspend
+                // while we're doing edits.
+                using (var readOnlyRegionEdit = _textBuffer.CreateReadOnlyRegionEdit())
+                {
+                    _readOnlyRegion = readOnlyRegionEdit.CreateDynamicReadOnlyRegion(
+                        _textBuffer.CurrentSnapshot.GetFullSpan(),
+                        SpanTrackingMode.EdgeInclusive,
+                        EdgeInsertionMode.Deny,
+                        callback: _ => !_updatingBuffer);
+
+                    readOnlyRegionEdit.Apply();
+                }
+
+                _workspace.WorkspaceChanged += OnWorkspaceChanged;
+
+                _batchingWorkQueue = new AsyncBatchingDelay(
+                    TimeSpan.FromSeconds(1),
+                    UpdateBufferContentsAsync,
+                    asyncListener: null,
+                    _cancellationTokenSource.Token);
+            }
+
+            public void Dispose()
+            {
+                using (var readOnlyRegionEdit = _textBuffer.CreateReadOnlyRegionEdit())
+                {
+                    readOnlyRegionEdit.RemoveReadOnlyRegion(_readOnlyRegion);
+                    readOnlyRegionEdit.Apply();
+                }
+
+                _workspace.WorkspaceChanged -= OnWorkspaceChanged;
+
+                // Cancel any remaining asynchronous work we may have had to update this file
+                _cancellationTokenSource.Cancel();
+            }
+
+            public async Task UpdateBufferContentsAsync(CancellationToken cancellationToken)
+            {
+                await TaskScheduler.Default;
```

Comment:
❔ Why switch to a background thread here? It certainly doesn't help the document open case since the call is placed in a `Run`, and `GetGeneratorDriverRunResultAsync` will already yield on the slow path.

Reply:
There's no guarantee that GetGeneratorDriverRunResultAsync would yield if there's only CPU bound work to do, no different than GetCompilationAsync or anything else. That said, you're right, the initial use is only going to make things _worse_ (since we're now jumping to the thread pool earlier), and I guess it's pointless for the async delay. Will remove.

Reply:
Jump removed.